x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,291 @@
+  private readonly wreck: typeof Wreck;
+  private readonly log: Logger;
+  private readonly config$: Observable<ProxyConfig>;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.nodeName = v4();
+    this.config$ = initializerContext.config.create<ProxyConfig>();
+    this.log = initializerContext.logger.get('proxy');
+    this.httpsAgent = new HTTPSAgent({ keepAlive: true });
+    this.httpAgent = new HTTPAgent({ keepAlive: true });
+    this.allowUnauthAgent = new HTTPAgent({ keepAlive: true });
+    this.wreck = Wreck.defaults({
+      agent: {
+        https: this.httpsAgent,
+        http: this.httpAgent,
+        httpsAllowUnauthorized: this.allowUnauthAgent,
+      },
+    });
+  }
+
+  public async setup(core: CoreSetup) {
+    // initialize cluster information
+    // we need to get this kibana instance's ip address
+    // it's hostname and determine which port to listen on
+    // should liveness timing be part of configuration?o
+    // mikhael -- talk about changing http service to allow multiple services
+    this.configSubscription = this.config$.subscribe(config => {
+      this.setConfig(config);
+    });
+
+    const config = await this.config$.pipe(first()).toPromise();
+    this.setConfig(config);
+    this.elasticsearch = core.elasticsearch.dataClient$;
+    this.proxyServer = await core.http.createNewServer({ port: config.port });
+  }
+
+  private setConfig(config: ProxyConfig) {
+    this.updateInterval = config.updateInterval || this.updateInterval;
+    this.timeoutThreshold = config.timeoutThreshold || this.timeoutThreshold;
+    this.port = config.port || this.port;
+  }
+
+  public async start(core: CoreStart) {
+    // start up an http service
+    // update es doc to explain that we exist
+    // start updateLivenessMap
+    this.mainLoop();
+  }
+
+  public async stop() {
+    // stop http service
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+
+    const nodes = await this.getNodeList();
+    nodes.delete(this.nodeName);
+    await this.updateNodeList(nodes);
+
+    if (this.configSubscription === undefined) {
+      return;
+    }
+
+    this.configSubscription.unsubscribe();
+    this.configSubscription = undefined;
+  }
+
+  public async assignResource(resource: string, data: RoutingNode): Promise<void> {
+    this.routingTable.set(resource, data);
+    const nodes = await this.getNodeList();
+    const currentTime = new Date().getTime();
+    await this.updateNodeList(this.updateLocalNode(nodes, currentTime));
+  }
+
+  public async unassignResource(resource: string) {
+    this.routingTable.delete(resource);
+    const nodes = await this.getNodeList();
+    const currentTime = new Date().getTime();
+    await this.updateNodeList(this.updateLocalNode(nodes, currentTime));
+  }
+
+  public async registerRouter(router: Router) {
+    this.proxyServer!.registerRouter(router);
+  }
+
+  public async proxyResource(resource: string) {
```

Comment:
what's proposed to be a public api? not sure we want to expose `hapi` as a part of public contract. end users interact with  [KibanaRequest](https://github.com/restrry/kibana/blob/f1a1a3dfbe2d19e241a20b6f71ef98f0fb999e7c/src/core/server/http/router/request.ts#L28)

Reply:
This is exactly the information I was hoping to get! I think KibanaRequest makes a ton of sense

Reply:
one note: that comment was related to the New Platform, as I assumed you develop plugin for it.

Reply:
this is for the new platform so this makes sense